eZ/Publish/Core/FieldType/Tests/FieldTypeTest.php

```diff
@@ -974,5 +979,20 @@ protected function doValidate($fieldDefinitionData, $value)
         return $validationErrors;
     }
 
+    /**
+     * @dataProvider provideInputForValuesEqual
+     */
+    public function testValuesEqual($inputValue1, $inputValue2)
```

Comment:
CS
```suggestion
    public function testValuesEqual(SPIValue $inputValue1, SPIValue $inputValue2): void
```

Reply:
we cannot declare in the SPIValue test because we would have to create a provider for it for all Type, currently, the second argument is an argument for FromHash to generate Value

Reply:
> we cannot declare in the SPIValue test because we would have to create a provider for it for all Type, currently, the second argument is an argument for FromHash to generate Value

Wow, this architecture is even more messy than I thought. It's not your fault of course, if the existing code were clean we'd move a lot faster.

AFAICS we can do:
```php
    /**
     * @dataProvider provideInputForValuesEqual
     *
     * @param mixed $inputValue1Hash
     */
    public function testValuesEqual($inputValue1Hash, SPIValue $inputValue2): void
```
because the 1st argument is actually mixed in this case, which is design flaw IMHO.